protocol/mqtt_paho/v2/protocol.go

```diff
@@ -0,0 +1,134 @@
+package mqtt_paho
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+	"github.com/eclipse/paho.golang/paho"
+
+	cecontext "github.com/cloudevents/sdk-go/v2/context"
+)
+
+type Protocol struct {
+	client         *paho.Client
+	connConfig     *paho.Connect
+	senderTopic    string
+	receiverTopics []string
+	qos            byte
+	retained       bool
+
+	// receiver
+	incoming chan *paho.Publish
+	// inOpen
+	openerMutex sync.Mutex
+}
+
+var (
+	_ protocol.Sender   = (*Protocol)(nil)
+	_ protocol.Opener   = (*Protocol)(nil)
+	_ protocol.Receiver = (*Protocol)(nil)
+	_ protocol.Closer   = (*Protocol)(nil)
+)
+
+func New(ctx context.Context, clientConfig *paho.ClientConfig, connConfig *paho.Connect, SenderTopic string,
+	ReceiverTopics []string, qos byte, retained bool,
+) (*Protocol, error) {
+	client := paho.NewClient(*clientConfig)
+	ca, err := client.Connect(ctx, connConfig)
+	if err != nil {
+		return nil, err
+	}
+	if ca.ReasonCode != 0 {
+		return nil, fmt.Errorf("failed to connect to %s : %d - %s", client.Conn.RemoteAddr(), ca.ReasonCode,
+			ca.Properties.ReasonString)
+	}
+
+	return &Protocol{
+		client:         client,
+		connConfig:     connConfig,
+		senderTopic:    SenderTopic,
+		receiverTopics: ReceiverTopics,
+		qos:            qos,
+		retained:       retained,
+		incoming:       make(chan *paho.Publish),
+		openerMutex:    sync.Mutex{},
+	}, nil
+}
+
+func (t *Protocol) Send(ctx context.Context, m binding.Message, transformers ...binding.Transformer) error {
+	var err error
+	defer m.Finish(err)
+
+	topic := cecontext.TopicFrom(ctx)
+	if topic == "" {
+		topic = t.senderTopic
+	}
+
+	msg := &paho.Publish{
+		QoS:    t.qos,
+		Retain: t.retained,
+		Topic:  topic,
+	}
+	err = WritePubMessage(ctx, m, msg, transformers...)
+	if err != nil {
+		return err
+	}
+
+	_, err = t.client.Publish(ctx, msg)
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func (t *Protocol) OpenInbound(ctx context.Context) error {
+	t.openerMutex.Lock()
```

Comment:
why need lock here?

Reply:
This method is to bootstrap a `Receiver/Responder`, it's a blocking call and ctx is used to stop the Inbound message Receiver/Responder. I think that can only be started once, that's the reason a lock is placed here. 

Do I understand correctly? @duglin  If not, please help me to clarify that. Thanks!

Reply:
Yes, other protocol implementations (besides AMQP IIUC) implement 

```go
// Opener is the common interface for things that need to be opened.
type Opener interface {
	// OpenInbound is a blocking call and ctx is used to stop the Inbound message Receiver/Responder.
	// Closing the context won't close the Receiver/Responder, aka it won't invoke Close(ctx).
	OpenInbound(ctx context.Context) error
}
```

in the same way using a mutex. It's a bit confusing as blocking can mean many things, in this project it seems to actually mimic a singleton.